homeassistant/components/emulated_kasa/__init__.py

```diff
@@ -0,0 +1,123 @@
+
+CONFIG_ENTITY_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_NAME): cv.string,
+        vol.Optional(CONF_POWER): vol.Any(
+            vol.Coerce(float),
+            cv.template,
+        ),
+    }
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
+            {
+                vol.Required(CONF_ENTITIES): vol.Schema(
+                    {cv.entity_id: CONFIG_ENTITY_SCHEMA}
+                ),
+            }
+        )
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the emulated_kasa component."""
+    conf = config.get(DOMAIN)
+    if not conf:
+        return True
+    entity_configs = conf[CONF_ENTITIES]
+
+    def devices():
+        """Devices to be emulated."""
+        yield from get_plug_devices(hass, entity_configs)
+
+    server = SenseLink(devices)
+
+    async def stop_emulated_kasa(event):
+        await server.stop()
+
+    async def start_emulated_kasa(event):
+        validate_configs(hass, entity_configs)
+        try:
+            await server.start()
+        except OSError as error:
+            _LOGGER.error("Failed to create UDP server at port 9999: %s", error)
+        else:
+            hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, stop_emulated_kasa)
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STARTED, start_emulated_kasa)
+
+    return True
+
+
+def validate_configs(hass, entity_configs):
+    """Validate that entities exist and ensure templates are ready to use."""
+    for entity_id, entity_config in entity_configs.items():
+        state = hass.states.get(entity_id)
+        if state is None:
+            _LOGGER.debug("Entity not found: %s", entity_id)
+            continue
+        if CONF_POWER in entity_config:
+            power_val = entity_config[CONF_POWER]
+            if isinstance(power_val, str) and is_template_string(power_val):
+                entity_config[CONF_POWER] = Template(power_val, hass)
+            elif isinstance(power_val, Template):
+                entity_config[CONF_POWER].hass = hass
+        elif state.domain == SENSOR_DOMAIN:
+            continue
+        elif ATTR_CURRENT_POWER_W in state.attributes:
+            continue
+        else:
+            _LOGGER.debug("No power value defined for: %s", entity_id)
+
+
+def get_plug_devices(hass, entity_configs):
+    """Produce list of plug devices from config entities."""
+    for entity_id, entity_config in entity_configs.items():
+        state = hass.states.get(entity_id)
+        if state is None:
+            continue
+        name = entity_config.get(CONF_NAME, state.name)
+
+        if state.state == STATE_ON or state.domain == SENSOR_DOMAIN:
+            if CONF_POWER in entity_config:
+                power_val = entity_config[CONF_POWER]
+                if isinstance(power_val, (float, int)):
+                    power = float(power_val)
+                elif isinstance(power_val, str):
```

Comment:
This seems like it might cause a bit of confusion since it takes a string or a template string and the string gets treated as an entity id.  I expect this will generate issues.  We should probably split this up into two separate options

Reply:
I was trying to keep it as simple as possible on the front end so people wouldn't have to worry about which option to use.  Do you think it could be a bug or just people would not know they are able to put either option as the power value?

Reply:
While it makes sense to me, its not particularly obvious to an outside observer that a string will get treated as an entity id, a number will be a power value, and that a template can also be used.    When I build schema I like to think about how successful the user will be when they don’t read the docs in detail or their first or second language is not English.    I think this might be particularly confusing for someone coming in fresh.